hanish/bot.py

```diff
@@ -0,0 +1,409 @@
+
+                # Check to see if anything has come through the channel
+                self.read_rtm_channel()
+
+                # Run any scheduled tasks
+                schedule.run_pending()
+
+                # Sleep for a bit until the next time we read the channel
+                time.sleep(POLLING_INTERVAL)
+
+        else:
+            raise SlackException(
+                "could not connect to API: invalid Slack API key or Bot ID?"
+            )
+
+        # If we've made it here, we've been shutdown
+        # TODO: convert to logging functionality
+        print("bot has been gracefully shutdown")
+
+    def stop(self):
+        """
+        Set the shutdown semaphore
+        """
+        self.shutdown = True
+
+    def read_rtm_channel(self):
+        """
+        Reads the RTM channel, parses output, filters messages based on type
+        and if they are directed at the bot. If so, passes those messages on
+        to the message handler command.
+        """
+
+        # Check for messages on the wire
+        # TODO: as soon as we connect, the last message comes through.
+        messages = self.slack.rtm_read()
+        if messages and len(messages) > 0:
+            # If there are messages, begin filtering and parsing.
+            for msg in messages:
+                # Filter only messages from other users
+                if msg["type"] == "message":
+                    # Check if an @botid appears in the message
+                    if self.atbotid.search(msg.get("text", "")):
+                        # Handle the atbotid message
+                        self.handle_message(msg)
+
+    def handle_message(self, msg):
+        """
+        Handles messages that are directed at the bot by parsing commands.
+        """
+        # Log the message received
+        # TODO: change to standardized logging functionality
+        print("[{}] {}".format(msg["ts"], msg["text"]))
+
+        # Track if an unknown command has occurred
+        unknown = True
+
+        # Determine if the message matches any commands
+        for name, command in self.commands.items():
+            match = command.search(msg["text"])
+            if match is not None:
+                unknown = False # We've found at least one command match.
+
+                # Get the method that handles this command
+                handler = getattr(self, "handle_{}_command".format(name), None)
+                if handler is None:
+                    raise NotImplementedError(
+                        "{} command not yet handled".format(name)
+                    )
+
+                try:
+                    # Handle the command
+                    handler(msg)
+                except HanishException as e:
+                    # Respond with hanish exceptions
+                    response = "Sorry, there was a problem with your request: {}".format(e)
+                    self.post(msg['channel'], response)
+                except Exception as e:
```

Comment:
i believe you gracefully handle exceptions that are used for user input validation, as well as programmatic failures with darksky, but i'm pretty sure any exceptions caused by interacting with the slack api will propagate (and presumably kill the process) 

can you talk briefly about the merits/disadvantages of using `raise/throw` as a program control mechanism (which is certainly a broad topic!)